prov/gni/src/gnix_datagram.c

```diff
@@ -334,41 +339,54 @@ int _gnix_dgram_bnd_post(struct gnix_datagram *d, gni_return_t *status_ptr)
 		goto err;
 	}
 
-	/*
-	 * if we get GNI_RC_ERROR_RESOURCE status return from
-	 * GNI_EpPostDataWId  that means that a previously posted wildcard
-	 * datagram has matched up with an incoming connect
-	 * request from the rank we are trying to send a connect
-	 * request to.  Don't treat this case as an error.
-	 */
-
 	fastlock_acquire(&d->nic->lock);
-	status = GNI_EpPostDataWId(d->gni_ep,
-				   d->dgram_in_buf,
-				   GNI_DATAGRAM_MAXSIZE,
-				   d->dgram_out_buf,
-				   GNI_DATAGRAM_MAXSIZE,
-				   (uint64_t)d);
-	fastlock_release(&d->nic->lock);
-	if ((status != GNI_RC_SUCCESS) &&
-		(status != GNI_RC_ERROR_RESOURCE)) {
+	if (d->pre_post_clbk_fn != NULL)
+		ret = d->pre_post_clbk_fn(d, &post);
+		if (ret != FI_SUCCESS)
 			GNIX_WARN(FI_LOG_EP_CTRL,
-			    "GNI_EpBind returned %s\n",
-			     gni_err_str[status]);
-			ret = gnixu_to_fi_errno(status);
-			goto err;
-	}
+				"pre_post_callback_fn: %d\n",
+				ret);
 
-	if (status_ptr != NULL)
-		*status_ptr = status;
+	if (post) {
+		/*
+		 * if we get GNI_RC_ERROR_RESOURCE status return from
+		 * GNI_EpPostDataWId  that means that a previously posted
+		 * wildcard datagram has matched up with an incoming
+		 * connect request from the rank we are trying to send
+		 * a connect request to.  Don't treat this case as an
+		 * error.
+		 */
+		status = GNI_EpPostDataWId(d->gni_ep,
+					   d->dgram_in_buf,
+					   GNI_DATAGRAM_MAXSIZE,
+					   d->dgram_out_buf,
+					   GNI_DATAGRAM_MAXSIZE,
+					   (uint64_t)d);
+		if (d->post_post_clbk_fn != NULL)
+			ret = d->post_post_clbk_fn(d, status);
+			if (ret != FI_SUCCESS)
+				GNIX_WARN(FI_LOG_EP_CTRL,
+				"post_post_callback_fn: %d\n",
+				ret);
+	}
+	fastlock_release(&d->nic->lock);
+	if (post) {
+		if ((status != GNI_RC_SUCCESS) &&
+			(status != GNI_RC_ERROR_RESOURCE)) {
+				GNIX_WARN(FI_LOG_EP_CTRL,
+				    "GNI_EpPostDataWId returned %s\n",
+				     gni_err_str[status]);
+				ret = gnixu_to_fi_errno(status);
+				goto err;
+		}
 
-	/*
-	 * datagram is active now, connecting
-	 */
-	if (status == GNI_RC_SUCCESS)
-		d->state = GNIX_DGRAM_STATE_CONNECTING;
-	else if (status == GNI_RC_ERROR_RESOURCE)
-		d->state = GNIX_DGRAM_STATE_ALREADY_CONNECTING;
+		/*
+		 * datagram is active now, connecting
+		 */
+		if ((status == GNI_RC_SUCCESS) ||
```

Comment:
Is this conditional necessary?  Isn't it always true, given the `goto err` above?